src/XMakeTasks/Exec.cs

```diff
@@ -561,19 +580,72 @@ protected override MessageImportance StandardOutputLoggingImportance
         private static readonly Encoding s_utf8WithoutBom = new UTF8Encoding(false);
 
         /// <summary>
-        /// Get encoding based on OS. This will fall back to previous behavior on any OS before Windows 7.
-        /// If the OS is greater than or equal to Windows 7, UTF8 encoding will be used for the cmd file.
+        /// Find the encoding for the batch file.
         /// </summary>
-        /// <remarks>UTF8 w/o BOM is used because cmd.exe does not like a BOM in a .cmd file.</remarks>
-        /// <returns>Encoding to use</returns>
-        private Encoding GetEncodingWithOsFallback()
+        /// <remarks>
+        /// The "best" encoding is the current OEM encoding, unless it's not capable of representing
+        /// the characters we plan to put in the file. If it isn't, we can fall back to UTF-8.
+        ///
+        /// Why not always UTF-8? Because tools don't always handle it well. See
+        /// https://github.com/Microsoft/msbuild/issues/397
+        /// </remarks>
+        private Encoding BatchFileEncoding()
         {
-            // Windows 7 (6.1) or greater
+            var defaultEncoding = EncodingUtilities.CurrentSystemOemEncoding;
+            string useUtf8 = string.IsNullOrEmpty(UseUtfEncoding) ? UseUtf8Detect : UseUtfEncoding;
+
+            // UTF8 is only supposed in Windows 7 (6.1) or greater.
             var windows7 = new Version(6, 1);
 
-            return Environment.OSVersion.Version >= windows7
-                ? s_utf8WithoutBom
-                : EncodingUtilities.CurrentSystemOemEncoding;
+            if (Environment.OSVersion.Version < windows7)
+            {
+                useUtf8 = UseUtf8Never;
+            }
+
+            switch (useUtf8.ToLowerInvariant())
```

Comment:
Probably doesn't matter for this case since we're only looking at 3 strings, but isn't the usual advice to uppercase things?